### PR TITLE
[core] Enable file index for DV table

### PR DESCRIPTION
**Pull request by leaves12138**
### Purpose

File index is only support Append Only tables yet. But it is able to support primary-key table like Deletion Vector Table.

This pull request, enable `File Index` in DV table.

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java [hunk 1]
```diff
@@ -121,10 +134,33 @@ protected boolean filterByStats(ManifestEntry entry) {
         }
 
         return filter.test(
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java:136]: Add "long schemaId = entry.file().schemaId();" in KeyValueFileStoreScan#filterByStats, and then use schemaId in KeyValueFileStoreScan#filterByStats and KeyValueFileStoreScan#testFileIndex.

File: paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java [hunk 2]
```diff
@@ -121,10 +134,33 @@ protected boolean filterByStats(ManifestEntry entry) {
...
-                serializer.evolution(stats.maxValues()),
-                serializer.evolution(stats.nullCounts(), entry.file().rowCount()));
+                        entry.file().rowCount(),
+                        serializer.evolution(stats.minValues()),
+                        serializer.evolution(stats.maxValues()),
+                        serializer.evolution(stats.nullCounts(), entry.file().rowCount()))
+                && (!fileIndexReadEnabled || testFileIndex(entry.file().embeddedIndex(), entry));
+    }
+
+    private boolean testFileIndex(@Nullable byte[] embeddedIndexBytes, ManifestEntry entry) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java:144]: filterByFileIndex?

File: paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java [hunk 3]
```diff
@@ -121,10 +134,33 @@ protected boolean filterByStats(ManifestEntry entry) {
...
+                        entry.file().schemaId(),
+                        id ->
+                                fieldValueStatsConverters.convertFilter(
+                                        entry.file().schemaId(), valueFilter));
+
+        try (FileIndexPredicate predicate =
+                new FileIndexPredicate(embeddedIndexBytes, dataRowType)) {
+            return predicate.testPredicate(dataPredicate);
+        } catch (IOException e) {
+            throw new RuntimeException("Exception happens while checking predicate.", e);
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java:162]: Exception happens while checking fileIndex predicate.

File: paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java [hunk 4]
```diff
@@ -121,10 +134,33 @@ protected boolean filterByStats(ManifestEntry entry) {
...
+                        entry.file().rowCount(),
+                        serializer.evolution(stats.minValues()),
+                        serializer.evolution(stats.maxValues()),
+                        serializer.evolution(stats.nullCounts(), entry.file().rowCount()))
+                && (!fileIndexReadEnabled || testFileIndex(entry.file().embeddedIndex(), entry));
+    }
+
+    private boolean testFileIndex(@Nullable byte[] embeddedIndexBytes, ManifestEntry entry) {
+        if (embeddedIndexBytes == null) {
+            return true;
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/operation/KeyValueFileStoreScan.java:146]: If fileIndex does not exist in embeddedIndex but in a separate file, what logic of filter by FileIndex in pk table with DV?

File: paimon-core/src/main/java/org/apache/paimon/KeyValueFileStore.java [hunk 5]
```diff
@@ -238,7 +238,8 @@ private KeyValueFileStoreScan newScan(boolean forWrite) {
                 options.scanManifestParallelism(),
                 options.deletionVectorsEnabled(),
                 options.mergeEngine(),
-                options.changelogProducer());
+                options.changelogProducer(),
+                options.fileIndexReadEnabled() && options.deletionVectorsEnabled());
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/KeyValueFileStore.java:242]: Add "File index support Append Only tables and pk table with DV" in [fileIndex doc](https://paimon.apache.org/docs/master/concepts/spec/fileindex/)

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**